### ISSUE TITLE
Red bar at top 

### ISSUE THREAD

**Issue by sonusainier** [nanoscopic/ios_remote_provider#35]
can we remove that red bar coming at top of screen??

**Comment by nanoscopic** [comment]
I'm generally opposed to removing the red bar, because there is no valid reason to do so.
The red bar is there to indicate to the user that their screen is being recorded/cast. That is exactly what is happening when a device is connected to the system and being controlled remotely.
The signal, intended by Apple, is being communicated properly.

Some potential reasons one could want to remove the red bar:

1. "Because I don't like seeing the red bar". What? This is ****. Come on now. Give me a real reason.

2. "Because our customers don't want to see the red bar". This is a slightly better reason, but isn't actually any better than 1. It's still not a reason.

3. "Because competing products don't need to show a red bar to show video". So? Who cares. Why does is matter? This is still not an actual reason to not have the red bar.

4. "Because it's unsightly". Alright, finally an actual reason. Still invalid. It's just a red bar. It's no big deal. If you think a red bar matters on style ****.

5. "Because my boss told me he doesn't want a red bar there." I don't care.

6. "Just because. Please remove it." Not a reason. ****

7. "Because it makes the customers wonder what the red bar means." Alright. This is potentially a legitimate reason. Still, the appropriate solution is to educate customers on what the red bar means. Hiding the red bar is, in my opinion, silly.

8. "Because I, the person asking, have influence over you, because if you do what I want, potentially my company will give you money." You'd best change your line of reasoning. I don't take kindly to anyone having me spend effort for no reason. Money is not my motivating factor.

In closing: Give me a **** reason to remove the red bar or ****.

Stars to protect the little children who can't handle adult language, who are apparently abundant in the technical world.

Some additional information about the red bar that I have shared previously but I will reiterate here:

1. I cannot remove or disable the red bar on the iOS device itself. It is enforced by Apple. Go bug Apple. They'll just ignore you. At least I am responding to this ridiculous request.

2. I can, and likely still will, add some code to the system that just removes the red color to some extent to address the "the red bar is unpleasant" reason. I still think this is a silly reason, but someday, on a blue moon, when I'm feeling particularly generous, I'll probably do it anyway. Even though it is ****.

3. There are other ways besides using the "Upload broadcast extension" / ReplayKit to stream video. One is by using the "Quicktime media path" using the hacky method to expose the screen stream to the normal media path in MacOS. This method works and I had it previously as the way video worked. I stopped using this method because it requires transcoding on the MacOS side which making scaling ( multiple devices per host ) very limited. You can also use the MJPEG stream from WDA. I am also in the future going to add this as an option, but the performance is worse through this method than the app method. I don't see the point of making performance worse just to avoid a red bar. Notably I believe most competitors are using some variant of the WDA MJPEG method for video. There is also the method of using wireless screen casting; but that also shows an indicator that that is occurring. One could also use HDMI dongle and capture the HDMI video stream... All these methods have been considered. The app method is the most reliable and the most supported by Apple themselves.

In closing, I am aware that people don't want the red bar there, but I have yet to hear a single reasonable explanation for why. My best guess as to the main reason is because "competing products don't have a red bar and we want to offer an equivalent service to them". My answer to that is "Yeah; they don't have a red bar, because they are doing it wrong."

**Comment by nanoscopic** [comment]
Discussed this more with a user of this project. I do agree that users will not know what the red bar means and will ask about it. That is tedious and would be nice to avoid. This just isn't a high priority issue as there are many other critical things at this time.

Anything that can be done to improve the experience for users I do agree with, it just has to be scheduled appropriately in the series of things that must be done. This is not high on the list at the moment.

Despite this, many people ask me about this, hence my frustration. I wish users would ask about the critical features and problems that exist rather than this.

**Comment by chewbacca70** [comment]
The only problem I see with the red bar is not the colour, but the possibility to click it and then stop the video stream and to stop the session, by accident.
But it is surely not the highest priority and can be avoided by teaching the user not to do so 😉

**Comment by nanoscopic** [comment]
@chewbacca70 Thanks for saying this. I am aware of that, and I agree that is a very important issue.

Clicking the bar halts the video and it has to then be restarted ( and right now it isn't restarted automatically if you do that )

Preventing clicking of the bar is easy enough to do and I should just get it done, as I agree it's an issue.

**Comment by nanoscopic** [comment]
I believe ultimately that the way to resolve this request is to use qvh method for those who can't stand the red bar.

qvh: github.com/danielpaulus/quicktime_video_hack/

It requires more provider machine computation power to decode mp4 and re-encode to jpeg to send to the browser, but it is generally the next best way to get video if not using the upload broadcast extension.

I believe this is the method uses by competing products as well in order to provide video without the red bar.

I do still believe that using the upload broadcast extension app is the best method, but I am all for users being able to decide for yourselves what you want.

To use qvh the best way is going to be using a raspberry pi for each and every connected device. The reason is because a raspberry pi is cheap and able to hardware decode the stream from the phone, and then hardware encode jpegs going back out.

I am investigating making this happen but it will take some more time and cooperation with @danielpaulus

**Comment by danielpaulus** [comment]
Or using xctest to record screenshots and stream as mjpeg, similar to how WDA does it. That works pretty well and lets the device do the heavy scaling operations :-)

**Comment by nanoscopic** [comment]
@danielpaulus I have been considering that time and again. I'll do that as well at some point. I want to first split off from WDA entirely and write the CFAgent I have been planning. CF doesn't actually use the normal WDA functions any more and is using custom functions I added into my fork of WDA. After some brief discussion with WDA maintainers I decided to move away from using WDA directly when I have the time.

What I dislike about the on-device path is that it pulls the image at full screen resolution and then has to resize those images after the call. What is worse is that it pulls the image in already encoded format ( png or jpeg ) and so it has to decode and re-encode to jpeg as well. For this reason the upload broadcast extension is better as it is handed the screen in raw format so there those additional decode/encode steps on device can be avoided. Essentially, using WDA for images is *worse* than using the upload broadcast extension. It does, I agree, not have the red bar at the top of the screen though.

The qvh method is great because it is using stable / reliable / optimized Apple code to encode the screen into h264, and is also doing it at a reasonable resolution size. Because it is optimized by Apple it uses less CPU time on the device than other methods ( app extension or WDA ) Some users are very picky about not having any extra CPU use as they want to be able to measure battery drain of their apps. I do personally think that the app extension CPU usage is quite low and shouldn't be a concern, but not all users agree with me.

In summary, yes, I will be adding support for all the various possible methods of attaining "video".

**Comment by nanoscopic** [comment]
For those who don't want the red bar, the default options no longer have it.

The new default setting for video is to use CFAgent to get frames. Doing it that way has no red bar. For those who want no red bar, that is the best option. Closing as this is a reasonable solution and is what competitors are doing imo to get video without a red bar.
